web-common/src/features/metrics-views/workspace/editor/MetricsEditor.svelte

```diff
@@ -1,51 +1,66 @@
   import { setLineStatuses } from "@rilldata/web-common/components/editor/line-status";
   import type { V1ParseError } from "@rilldata/web-common/runtime-client";
   import { yamlSchema } from "codemirror-json-schema/yaml";
   import { mapParseErrorsToLines } from "../../errors";
   import MetricsEditorContainer from "./MetricsEditorContainer.svelte";
   import { createPlaceholder } from "./create-placeholder";
   import metricsSchema from "./metrics-schema.json";
-  import { createUpdateMetricsCallback } from "./update-metrics";
-
   import { JSONSchema7 } from "json-schema";
+  import Editor from "@rilldata/web-common/features/editor/Editor.svelte";
+  import { FileArtifact } from "@rilldata/web-common/features/entity-management/file-artifacts";
+  import { metricsExplorerStore } from "@rilldata/web-common/features/dashboards/stores/dashboard-stores";
+  import { createPersistentDashboardStore } from "@rilldata/web-common/features/dashboards/stores/persistent-dashboard-state";
+  import { FileExtensionToEditorExtension } from "@rilldata/web-common/features/editor/getExtensionsForFile";
 
   export let filePath: string;
-  export let yaml: string;
   export let metricViewName: string;
   export let allErrors: V1ParseError[];
+  export let fileArtifact: FileArtifact;
+  export let autoSave: boolean;
+
+  $: ({ remoteContent } = fileArtifact);
 
-  let view: EditorView;
+  let editor: EditorView;
   const metricsJsonSchema = metricsSchema as JSONSchema7;
 
   /** create placeholder codemirror extension and set the inner svelte component
    * generated by the createPlaceholder callback.
    */
   $: placeholderElements = createPlaceholder(filePath, metricViewName);
-  $: if (view) placeholderElements.component.setEditorView(view);
+  $: if (editor) placeholderElements.component.setEditorView(editor);
 
-  /** create an updateMetrics event callback based on the queryClient
-   * and metricsDefName.
-   */
-  $: updateMetrics = createUpdateMetricsCallback(filePath, metricViewName);
-
-  $: lineBasedRuntimeErrors = mapParseErrorsToLines(allErrors, yaml);
+  $: lineBasedRuntimeErrors = mapParseErrorsToLines(
+    allErrors,
+    $remoteContent ?? "",
+  );
   /** display the main error (the first in this array) at the bottom */
   $: mainError = lineBasedRuntimeErrors?.at(0);
 
   /** If the errors change, run the following transaction. */
-  $: if (view) setLineStatuses(lineBasedRuntimeErrors, view);
+  $: if (editor) setLineStatuses(lineBasedRuntimeErrors, editor);
 </script>
 
-<MetricsEditorContainer error={yaml?.length ? mainError : undefined}>
-  <YAMLEditor
-    key={filePath}
-    bind:view
-    content={yaml}
-    extensions={[placeholderElements.extension, yamlSchema(metricsJsonSchema)]}
-    whenFocused
-    on:save={updateMetrics}
+<MetricsEditorContainer error={$remoteContent ? mainError : undefined}>
+  <Editor
+    bind:autoSave
+    bind:editor
+    onSave={(content) => {
```

Comment:
**Issue**: The onSave callback function does not handle asynchronous operations properly, which can lead to race conditions or unhandled promise rejections if the operations inside it fail. <br> **Fix**: Modify the onSave callback to properly handle asynchronous operations using async/await to ensure that errors are caught and handled properly. <br> **Code Suggestion**: 
 ```
 @@ -49,7 +49,11 @@
 onSave={(content) => {
 -  metricsExplorerStore.remove(metricViewName);
 -  createPersistentDashboardStore(metricViewName).reset();
 +  (async () => {
 +    try {
 +      await metricsExplorerStore.remove(metricViewName);
 +      await createPersistentDashboardStore(metricViewName).reset();
 +    } catch (error) {
 +      console.error('Failed to save metrics:', error);
 +    }
 +  })();
   if (!content?.length) {
     setLineStatuses([], editor);
   }
 }}
 ```

Reply:
These functions are not async.